File: ingest/corrections.py
"""The corrections layer, SCHEMA_VERSION 2.

Doc 0, three times over:

    "Corrections never overwrite model output."
    "events and tracks store raw model output only."
    "Keeping both is the whole point: overwriting destroys the ability to measure whether
     the model is improving."

Corrections are stored as their own rows and composed onto raw records at read time.
`raw=true` skips this entirely and returns what the model actually said, which is what the
accuracy comparison and the training-data export need.

If this module is ever bypassed and an event row is UPDATEd in place, the original prediction
is gone permanently -- a correction records the new value, not the old one.
"""

from .serializers import event_to_dict, track_to_dict


def _sorted(corrections):
    return sorted(corrections, key=lambda c: (c.created_at is None, c.created_at))


def apply_corrections(events, corrections) -> list[dict]:
    """Compose corrections onto raw events. Both arguments are ORM rows.

    A TRACK-scoped correction re-attributes the track and every event on it, as one action.
    Doc 3 calls that the primary correction path: one bad OCR read mislabels forty-odd events
    and every box on that robot.
    """
    by_id: dict[str, dict] = {e.event_id: event_to_dict(e) for e in events}
    deleted: set[str] = set()

    for correction in _sorted(corrections):
        action = correction.action
        target = correction.target_id

        if correction.scope == "track":
            try:
                track_id = int(target)
            except (TypeError, ValueError):
                continue
            patch = correction.fields or {}
            for row in by_id.values():
                if row.get("track_id") != track_id:
                    continue
                row.update(patch)
                row["corrected"] = True
                row["correction_id"] = correction.correction_id
            continue

        if action == "delete":
            deleted.add(target)
            continue

        if action == "create":
            fields = dict(correction.fields or {})
            fields["event_id"] = target
            fields.setdefault("schema_version", 2)
            fields.setdefault("source", "manual")
            fields["corrected"] = True
            fields["correction_id"] = correction.correction_id
            by_id[target] = fields
            deleted.discard(target)
            continue

        if action == "edit":
            existing = by_id.get(target)
            if existing is None or not correction.fields:
                # An edit against an event that no longer exists is a no-op, not an error.
                continue
            by_id[target] = {
                **existing,
                **correction.fields,
                "corrected": True,
                "correction_id": correction.correction_id,
            }

    rows = [row for event_id, row in by_id.items() if event_id not in deleted]
    rows.sort(key=lambda r: r.get("t_seconds") or 0.0)
    return rows


def apply_track_corrections(tracks, corrections) -> list[dict]:
    """Compose track-scoped corrections onto the tracks themselves.

    Without this the overlay keeps the old bumper label even though every event on the track
    was re-attributed -- the boxes read `team` from the track, not from events.
    """
    rows = {t.track_id: track_to_dict(t) for t in tracks}
    for correction in _sorted(corrections):
        if correction.scope != "track" or not correction.fields:
            continue
        try:
            track_id = int(correction.target_id)
        except (TypeError, ValueError):
            continue
        row = rows.get(track_id)
        if row is None:
            continue
        for key in ("team", "alliance", "team_confidence"):
            if key in correction.fields:
                row[key] = correction.fields[key]
    return list(rows.values())

File: web/src/lib/corrections.ts
// The corrections layer.
//
// Doc 0: "Corrections never overwrite model output... Keeping both is the whole point:
// overwriting destroys the ability to measure whether the model is improving."
//
// v2 moved correction *state* onto the API response: every returned event carries `corrected`
// and `correction_id`, so the client-side diff that used to reconstruct it is gone. What
// remains is pairing the corrected view against raw so the inspector can still show what the
// model originally said, and applying corrections locally in fixture mode.

import type { Correction, ScoutEvent } from '../contracts';

export interface ViewEvent extends ScoutEvent {
  /** The uncorrected event, when this row was edited. Null for untouched or created rows. */
  original: ScoutEvent | null;
  /** Which fields differ from the model's original, for highlighting. */
  changedFields: ReadonlyArray<keyof ScoutEvent>;
}

const NEVER_COMPARE: ReadonlySet<string> = new Set(['corrected', 'correctionId']);

/**
 * Pair the corrected stream against raw. `corrected` comes from the API, not from this diff --
 * the diff only recovers the before-values for display.
 */
export function toViewEvents(corrected: ScoutEvent[], raw: ScoutEvent[]): ViewEvent[] {
  const rawById = new Map(raw.map((e) => [e.eventId, e]));
  return corrected
    .map((e) => {
      const before = rawById.get(e.eventId);
      if (!before) {
        // Present in the corrected view but not in raw: a human created it.
        return { ...e, original: null, changedFields: [] };
      }
      const changed = (Object.keys(e) as Array<keyof ScoutEvent>).filter(
        (k) => !NEVER_COMPARE.has(k as string) && e[k] !== before[k]
      );
      return { ...e, original: changed.length ? before : null, changedFields: changed };
    })
    .sort((a, b) => a.tSeconds - b.tSeconds);
}

/** Events present in raw but absent from the corrected view -- i.e. deleted by a human. */
export function deletedEvents(raw: ScoutEvent[], corrected: ScoutEvent[]): ScoutEvent[] {
  const keep = new Set(corrected.map((e) => e.eventId));
  return raw.filter((e) => !keep.has(e.eventId));
}

/**
 * Compose corrections onto raw events. Only the fixture client needs this -- against a real
 * backend, component 2 does it and `raw=true` returns the uncorrected stream.
 *
 * A track-scoped correction re-attributes the track and EVERY event on it, as one action.
 * Doc 3: that is the primary correction path, because one bad OCR read mislabels forty-odd
 * events and every box on that robot.
 */
export function applyCorrections(
  raw: ScoutEvent[],
  corrections: Correction[]
): ScoutEvent[] {
  const byId = new Map<string, ScoutEvent>(raw.map((e) => [e.eventId, { ...e }]));
  const deleted = new Set<string>();

  // Oldest first, so a later correction to the same target wins.
  const ordered = [...corrections].sort((a, b) => a.createdAt.localeCompare(b.createdAt));

  for (const c of ordered) {
    if (c.scope === 'track') {
      const trackId = Number(c.targetId);
      for (const [id, e] of byId) {
        if (e.trackId !== trackId) continue;
        byId.set(id, { ...e, ...c.fields, corrected: true, correctionId: c.correctionId });
      }
      continue;
    }

    if (c.action === 'delete') {
      deleted.add(c.targetId);
      continue;
    }

    if (c.action === 'create') {
      const f = c.fields ?? {};
      byId.set(c.targetId, {
        jobId: f.jobId ?? '',
        matchId: f.matchId ?? '',
        eventId: c.targetId,
        team: f.team ?? null,
        trackId: f.trackId ?? null,
        tSeconds: f.tSeconds ?? 0,
        phase: f.phase ?? 'unknown',
        eventType: f.eventType ?? 'shot_attempt',
        confidence: f.confidence ?? 1,
        fieldX: f.fieldX ?? null,
        fieldY: f.fieldY ?? null,
        goal: f.goal ?? null,
        source: 'manual',
        corrected: true,
        correctionId: c.correctionId,
      });
      deleted.delete(c.targetId);
      continue;
    }

    const target = byId.get(c.targetId);
    if (!target || !c.fields) continue;
    byId.set(c.targetId, {
      ...target,
      ...c.fields,
      corrected: true,
      correctionId: c.correctionId,
    });
  }

  return [...byId.entries()]
    .filter(([id]) => !deleted.has(id))
    .map(([, e]) => e)
    .sort((a, b) => a.tSeconds - b.tSeconds);
}

/** Apply track-scoped corrections to the tracks themselves, so the overlay labels follow. */
export function applyTrackCorrections<T extends { trackId: number; team: number | null }>(
  tracks: T[],
  corrections: Correction[]
): T[] {
  const byTrack = new Map<number, Partial<{ team: number | null }>>();
  for (const c of [...corrections].sort((a, b) => a.createdAt.localeCompare(b.createdAt))) {
    if (c.scope !== 'track' || !c.fields) continue;
    byTrack.set(Number(c.targetId), { team: c.fields.team ?? null });
  }
  return tracks.map((t) => {
    const patch = byTrack.get(t.trackId);
    return patch ? { ...t, ...patch } : t;
  });
}
